presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -103,21 +125,79 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
-                    node.getTable(),
-                    node.getOutputSymbols(),
-                    annotateColumnsWithSubfields(node.getAssignments()),
-                    node.getLayout(),
-                    node.getCurrentConstraint(),
-                    node.getEnforcedConstraint());
+            if (subfieldPaths.isEmpty()) {
+                return node;
+            }
+
+            Assignments.Builder projections = Assignments.builder();
+            ImmutableMap.Builder<Symbol, ColumnHandle> newAssignments = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Symbol> symbolMapBuilder = ImmutableMap.builder();
+            for (Map.Entry<Symbol, ColumnHandle> entry : node.getAssignments().entrySet()) {
+                if (fullColumnUses.contains(entry.getKey())) {
+                    newAssignments.put(entry);
+                    projections.putIdentity(entry.getKey());
+                    continue;
+                }
+
+                List<SubfieldPath> subfields = new ArrayList();
+                for (SubfieldPath path : subfieldPaths) {
+                    if (path.getColumnName().equals(entry.getKey().getName())) {
+                        subfields.add(path);
+                    }
+                }
+                if (subfields.isEmpty()) {
+                    newAssignments.put(entry);
+                    projections.putIdentity(entry.getKey());
+                    continue;
+                }
+                // Compute the leaf subfields. If we have a.b.c and
+                // a.b then a.b is the result. If a path is a prefix
+                // of another path, then the longer is discarded.
+                List<SubfieldPath> leafPaths = new ArrayList();
+                for (SubfieldPath path : subfields) {
+                    if (!prefixExists(path, subfields)) {
+                        leafPaths.add(path);
+                    }
+                }
+
+                Symbol newSymbol = symbolAllocator.newSymbol(entry.getKey());
+                projections.put(entry.getKey(),
+                        new FunctionCall(
+                                FILTER_BY_SUBSCRIPT_PATHS,
+                                ImmutableList.of(
+                                        newSymbol.toSymbolReference(),
+                                        new ArrayConstructor(leafPaths.stream()
+                                                .map(p -> new StringLiteral(p.getPath()))
+                                                .collect(toImmutableList())))));
+                newAssignments.put(newSymbol, entry.getValue());
+                symbolMapBuilder.put(entry.getKey(), newSymbol);
+            }
+
+            Map<Symbol, Symbol> symbolMap = symbolMapBuilder.build();
+            if (symbolMap.isEmpty()) {
+                return node;
+            }
+
+            ProjectNode projectNode = new ProjectNode(
+                    idAllocator.getNextId(),
+                    new TableScanNode(
```

Comment:
I came across a similar thing and thought it might be better to flatten.  While having this nested with indentation shows the layout, it can make it hard to see which parameter belongs to which node.  This is just my personal preference though.

Reply:
@tdcmeehan How about this?

```
            TableScanNode tableScanNode = new TableScanNode(
                    node.getId(),
                    node.getTable(),
                    node.getOutputSymbols().stream()
                            .map(s -> symbolMap.getOrDefault(s, s))
                            .collect(toImmutableList()),
                    newAssignments.build(),
                    node.getLayout(),
                    node.getCurrentConstraint(),
                    node.getEnforcedConstraint());

            return new ProjectNode(idAllocator.getNextId(), tableScanNode, projections.build());
```